keps/sig-apps/1847-autoremove-statefulset-pvcs/README.md

```diff
@@ -0,0 +1,508 @@
+are the precise PVCs meant when referring to the volume or PVC for Pod below, and these
+are the only PVCs modified with an ownerRef. Other PVCs referenced by the StatefulSet Pod
+template are not affected by this behavior.
+
+#### `DeleteOnScaledown`
+
+If `PersistentVolumeClaimDeletePolicy` is set to `DeleteOnScaledown`, the Pod will be set
+as the owner of the PVCs created from the `VolumeClaimTemplates` just before the
+scale-down is performed by the StatefulSet controller.  When a Pod is deleted, the PVC
+owned by the Pod is also deleted. When `DeleteOnScaledown` policy is set and the
+Statefulset gets deleted the PVCs also will get deleted (similar to
+`DeleteOnStatefulSetDeletion` policy).
+
+The current StatefulSet controller implementation ensures that the manually deleted pods
+are restored before the scale-down logic is run. This combined with the fact that the
+owner references are set only before the scale-down will ensure that manual deletions do
+not automatically delete the PVCs in question.
+
+During scale-up, if a PVC has an OwnerRef that does not match the Pod, it indicates that
+the PVC was referred to by the deleted Pod and is in the process of getting
+deleted. The controller will skip the reconcile loop until PVC deletion finishes, avoiding
+a race condition.
+
+In addition, on PVC creation an OwnerRef is added for when the StatefulSet is
+deleted. See the `DeleteOnStatefulSetDeletion` policy below for further details
+how this will be handled.
+
+#### `DeleteOnStatefulSetDeletion`
+
+When `PersistentVolumeClaimDeletePolicy` is set to
+`DeleteOnStatefulSetDeletion`, when a VolumeClaimTemplate PVC is created, an
+owner reference in PVC will be added to point to the StatefulSet. When a
+scale-up or scale-down occurs, the PVC is unchanged.  PVCs previously in use
+before scale-down will be used again when the scale-up occurs.
+
+In the existing StatefulSet reconcile loop, the associated VolumeClaimTemplate
+PVCs will be checked to see if the ownerRef is correct according to the
+`PersistentVolumeClaimDeletePolicy` and updated accordingly. This includes PVCs
+that have been manually provisioned. It will be most consistent and easy
+to reason about if all VolumeClaimTemplate PVCs are treated uniformly rather
+than trying to guess at their provenance.
+
+When the StatefulSet is deleted, these PVCs will also be deleted, but only after
+the Pod gets deleted. Since the Pod StatefulSet ownership has
+`blockOwnerDeletion` set to `true`, pods will get deleted before the StatefulSet
+is deleted. The `blockOwnerDeletion` for PVCs will be set to `false` which
+ensures that PVC deletion happens only after the StatefulSet is deleted. This is
+necessary because of PVC protection which does not allow PVC deletion until all
+pods referencing it are deleted.
+
+`Retain` `PersistentVolumeClaimDeletePolicy` will ensure the current behaviour: no PVC
+deletion is performed as part of StatefulSet controller.
+
```

Comment:
Suggest also discussing what happens when `PersistentVolumeClaimDeletePolicy` is changed. 

The controller would remove OwnerRefs from PVCs, but given that there can be multiple OwnerRefs in an object, the controller needs to make sure it removes the correct one (either the Pod or StatefulSet) instead of just setting it to `nil`. The Pod can be found by name matching, and the StatefulSet can be found by checking the Pod's controller ref, for example.

Reply:
Clarified.

Reply:
For pod case, you're gonna need to check both name and uid. The former is stable for sts, the latter always changes. That's how the GC will know which exactly pod owns the right PVC. I doesn't have to be explicitly called out here, but it will be critical in the controller code.